hazelcast/src/main/java/com/hazelcast/internal/probing/ProbeRegistryImpl.java

```diff
@@ -0,0 +1,681 @@
+            this.source = source;
+            this.update = reprobeFor(source.getClass());
+            if (update != null) {
+                ReprobeCycle reprobe = update.getAnnotation(ReprobeCycle.class);
+                updateIntervalMs = Probing.updateInterval(reprobe.value(), reprobe.unit());
+                updateLevel = reprobe.level();
+            } else {
+                updateIntervalMs = -1L;
+                updateLevel = ProbeLevel.DEBUG;
+            }
+        }
+
+        void updateIfNeeded() {
+            if (update != null) {
+                long next = nextUpdateTimeMs.get();
+                long now = Clock.currentTimeMillis();
+                if (now > next && nextUpdateTimeMs.compareAndSet(next, now + updateIntervalMs)) {
+                    try {
+                        update.invoke(source, EMPTY_ARGS);
+                    } catch (Exception e) {
+                        LOGGER.warning("Failed to update source: "
+                                + update.getDeclaringClass().getSimpleName() + "."
+                                + update.getName(), e);
+                    }
+                }
+            }
+        }
+
+        static Method reprobeFor(Class<?> type) {
+            for (Method m : type.getDeclaredMethods()) {
+                if (m.isAnnotationPresent(ReprobeCycle.class)) {
+                    m.setAccessible(true);
+                    return m;
+                }
+            }
+            return null;
+        }
+    }
+
+    private static final class ProbingCycleImpl
+    implements ProbingCycle, ProbingCycle.Tags, ProbeRenderContext {
+
+        private final StringBuilder tags = new StringBuilder(128);
+        private final Collection<ProbeSourceEntry> sources;
+        private CharSequence lastTagName;
+        private int lastTagValuePosition;
+
+        // render cycle state
+        private ProbeRenderer renderer;
+        private ProbeLevel level;
+
+        ProbingCycleImpl(Collection<ProbeSourceEntry> sources) {
+            this.sources = sources;
+        }
+
+        @Override
+        public void renderAt(ProbeLevel level, ProbeRenderer renderer) {
+            this.level = level;
+            this.renderer = renderer;
+            openContext(); // reset
+            for (ProbeSourceEntry entry : sources) {
+                if (isProbed(entry.updateLevel)) {
+                    entry.updateIfNeeded();
+                }
+                try {
+                    openContext(); // just to be sure it is done even if ommitted in the source
+                    entry.source.probeIn(this);
+                } catch (Exception e) {
+                    LOGGER.warning("Exception while probing source "
+                            + entry.source.getClass().getSimpleName(), e);
+                }
+            }
+        }
+
+        @Override
+        public void probe(final ProbeLevel level, Object instance, final String... methods) {
```

Comment:
varargs can cause litter. Probably this is not what we want.

Reply:
I'll make it an array.